### PR TITLE
refactor(web): osk inner-frame abstraction (help text vs std OSK) 🍕

### PR DESCRIPTION
**Pull request by jahorton**
Yet another piece of the overall OSK-Core refactor.

While rarely used, there are alternative "keyboard visualizations" that `OSKManager` supports.  A great example comes from desktop form-factor `sil_euro_latin`, which doesn't display a keyboard.  Instead, that keyboard prefers a "help page" visualization:

image

Among other things, this PR encapsulates the code needed to properly support this "help page" approach.  (If need be, those changes are contained pretty well within the first commit.)

Furthermore, as this "help page" takes the place of a `VisualKeyboard` - like, literally takes the place at which one would be inserted - this provides an abstract interface that both `VisualKeyboard` and the new `HelpPage` class both adhere to.  This interface then allows us to clean up and better consolidate how `OSKManager.Load` constructs its elements and layout.

## User Testing
@keymanapp/testers 

#### General Description

<details>
<summary>Plain-text description</summary>

Run the following tests, when possible, for the two following keyboards:
- (LATIN) `sil_euro_latin`.
- (KHMER) `khmer_angkor`.  (Selected as a "control" / "more normal" keyboard.)

Loading test (LOAD): As this PR is centered around how the OSK is constructed within KMW, simply make sure that the keyboard loads as expected on each different device type.

Collateral prevention (ROBUST): Test a few keystrokes for good measure; I don't expect any collateral effects here, but it's good to be sure.

1. `=` then `e`:  should emit `ə`
2. 'm': should emit `m`

Help text insertion (SHORTCUT):  Click the desktop form-factor's "help page" link for `ə` and verify that it is properly output.

image
- So, click the bottom-right part that displays the `ə`.

</details>

### Explicit test list

Run this suite on the following platform / browser combinations:

- [x] Windows / Firefox
- [x] Windows / Chrome
- [x] macOS / Safari
<details>
<summary><strong>Per-platform suite and template</strong> (desktop form-factor)</summary>

```
- [ ] `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- [ ] `LATIN.ROBUST.DESKTOP.1`:  physically type `=` then `e`.
    - Expected result: `ə`  
- [ ] `LATIN.ROBUST.DESKTOP.2`:  physically type `m`.
    - Expected result: `m`
- [ ] `LATIN.SHORTCUT`:  click on the "help-text OSK"'s `ə` button
    - Expected result: `ə`
    - Refer to the following image for the "button": 
        <img width="139" alt="image" src="https://user-images.githubusercontent.com/25213402/126125311-6c95a743-a87f-4bdf-bc24-090a3c449fa1.png">
- [ ] `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- [ ] `KHMER.ROBUST.1`:  physically type a `ស`.
```

- `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- `LATIN.ROBUST.DESKTOP.1`:  physically type `=` then `e`.
    - Expected result: `ə`  
- `LATIN.ROBUST.DESKTOP.2`:  physically type `m`.
    - Expected result: `m`
- `LATIN.SHORTCUT`:  click on the "help-text OSK"'s `ə` button
    - Expected result: `ə`
    - Refer to the following image for the "button": 
        image
- `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- `KHMER.ROBUST.1`:  physically type a `ស`.
</details>

Run _this_ suite on the following platform / browser combinations

- [x] macOS / Chrome emulation / iPhone X
- [x] Android / Keyman app

<details>
<summary><strong>Per-platform suite and template</strong> (mobile form-factors)</summary>

```
- [ ] `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- [ ] `LATIN.ROBUST.MOBILE.1`:  use the OSK to type `=` (numeric layer) then `e`.
    - Expected result: `=e`
- [ ] `LATIN.ROBUST.MOBILE.2`:  use a longpress on `e` to type `ə`.  (Second-to-last subkey)
    - Expected result: `ə`
- [ ] `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- [ ] `KHMER.ROBUST.1`:  use the OSK to type a `ស`.
```

- `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- `LATIN.ROBUST.MOBILE.1`:  use the OSK to type `=` (numeric layer) then `e`.
    - Expected result: `=e`
- `LATIN.ROBUST.MOBILE.2`:  use a longpress on `e` to type `ə`.  (Second-to-last subkey)
    - Expected result: `ə`
- `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- `KHMER.ROBUST.1`:  use the OSK to type a `ស`.

</details>

### CODE REVIEW

File: common/core/web/keyboard-processor/src/keyboards/keyboard.ts [hunk 1]
```diff
@@ -98,24 +98,27 @@ namespace com.keyman.keyboards {
...
-    get hasHelpHTML(): boolean {
+    get hasScript(): boolean {
       return !!this.scriptObject['KHF'];
     }
 
     /**
-     * Replaces the OSK with custom HTML, which may be interactive (like with sil_euro_latin).
+     * Embeds a custom script for use by the OSK, which may be interactive (like with sil_euro_latin).
+     * Note:  this must be called AFTER any contents of `helpText` have been inserted into the DOM.
+     * (See sil_euro_latin's source -> sil_euro_latin_js.txt)
```
Comment (jahorton) [inline common/core/web/keyboard-processor/src/keyboards/keyboard.ts:117]: This PR gave me a great opportunity to dive into `sil_euro_latin` and see how it _really_ ticks.

The `KHF` function wraps the keyboard's embedded _script_.

https://github.com/keymanapp/keyboards/blob/1c9365083c10acb758289c5ffb140f5fdc3a4526/release/sil/sil_euro_latin/source/sil_euro_latin.kmn#L15
```
store(&KMW_EMBEDJS) 'sil_euro_latin_js.txt'
```

https://github.com/keymanapp/keyboards/blob/1c9365083c10acb758289c5ffb140f5fdc3a4526/release/sil/sil_euro_latin/source/sil_euro_latin_js.txt#L1-L4

```
this.KHF=function(e) { 
  var h = document.getElementsByTagName('head');
  var stext = 
  "#keyboard_europeanlatin_help { text-align: left; padding: 0; background: #f5e3de; font-size: 8pt; font-family: Tahoma; cursor: default; padding: 4px; }"+
```

Sadly, it seems that Github doesn't like to expand cross-repo permalinks.  ☹️

Comment (jahorton) [reply common/core/web/keyboard-processor/src/keyboards/keyboard.ts:117]: Side-note:  that `KHF` definition has some _nasty_ stuff, some of which will make Shadow DOM tricky to perform for such a "help page" approach in the future.

Like, direct insertion of a style sheet to the <head> element, as seen here.

It also blatantly ignores the provided `e` parameter, preferring to directly retrieve the one it wants via `document.getElementById()` on an element defined in its help-text section.

File: web/source/kmwbase.ts [hunk 2]
```diff
@@ -685,7 +685,7 @@ namespace com.keyman {
 
       PKbd = PKbd || this.core.activeKeyboard;
 
-      return com.keyman.osk.VisualKeyboard.buildDocumentationKeyboard(PKbd, argFormFactor, argLayerId, this.osk);
+      return com.keyman.osk.VisualKeyboard.buildDocumentationKeyboard(PKbd, argFormFactor, argLayerId, this.osk.getKeyboardHeight());
```
Comment (jahorton) [inline web/source/kmwbase.ts:688]: Internal-only method; the external API that uses it remains unchanged.

File: web/source/kmwembedded.ts [hunk 3]
```diff
@@ -250,7 +250,11 @@ namespace com.keyman.text {
    * correctOSKTextSize handles rotation event -- currently rebuilds keyboard and adjusts font sizes
    */
   keymanweb['correctOSKTextSize']=function() {
-    if(osk && osk.vkbd && osk.vkbd.adjustHeights(osk)) {
+    let osk = keymanweb.osk;
+    if(osk && osk.vkbd && osk.vkbd.adjustHeights(osk.getKeyboardHeight())) {
+      var b: HTMLElement = osk._Box, bs=b.style;
+      bs.height=bs.maxHeight=osk.vkbd.computedAdjustedOskHeight(this.getHeight())+'px';
```
Comment (jahorton) [inline web/source/kmwembedded.ts:256]: The new lines of the inner block are relocated from within `VisualKeyboard.adjustHeights`, removing the need for that function to reference this class's `_Box` field.  (It's _very_ inconvenient for a few specific calls elsewhere.)

File: web/source/osk/oskManager.ts [hunk 4]
```diff
@@ -191,69 +189,54 @@ namespace com.keyman.osk {
...
-      if(innerFrame.id == 'keymanweb_title_bar') {
-        // Desktop order is title_bar, banner_container, inner-frame
-        innerFrame=<HTMLDivElement> innerFrame.nextSibling.nextSibling;
-      } else if (innerFrame.id == 'keymanweb_banner_container') {
-        innerFrame=<HTMLDivElement> innerFrame.nextSibling;
-      }
-      innerFrame.className = 'kmw-osk-inner-frame' + kbdClass;
+
+      var kbdClassSuffix = ' kmw-keyboard-' + kbdID;
+      kbdView.element.className = kbdView.element.className + kbdClassSuffix;
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:239]: `kbdView.element` is equivalent to the old `innerFrame` - but this time, we don't need to search `_Box`'s hierarchy to find it.

File: web/source/osk/visualKeyboard.ts [hunk 5]
```diff
@@ -1201,7 +1208,7 @@ namespace com.keyman.osk {
       return null;
     }
 
-    show(host: OSKManager) {
+    updateState() {
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:1211]: It never really was a "show" method... and not really even an "update layout" method.

It's also 100% internal, so the name-change is safe.

File: web/source/osk/oskManager.ts [hunk 6]
```diff
@@ -820,8 +807,17 @@ namespace com.keyman.osk {
       // TODO:  Move this into the VisualKeyboard class!
       // The following code will always be executed except for externally created OSK such as EuroLatin
       if(this.vkbd && this.vkbd.ddOSK) {
+        // Always adjust screen height if iPhone or iPod, to take account of viewport changes
+        // Do NOT condition upon form-factor; this line prevents a bug with displaying
+        // the predictive-text banner on the initial keyboard load.  (Issue #2907)
+        if(device.touchable && device.OS == 'iOS') {
+          this.vkbd.adjustHeights(this.getKeyboardHeight());
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:814]: This particular call (`adjustHeights`) is no longer included in `VisualKeyboard`'s `show` / `updateState` method.

File: web/source/osk/visualKeyboard.ts [hunk 7]
```diff
@@ -1260,24 +1259,21 @@ namespace com.keyman.osk {
         fs=fs/keyman.util.getViewportScale();
       }
 
-      let paddedHeight = this.computedAdjustedOskHeight(host.getKeyboardHeight());
-
-      var b: HTMLElement = _Box, bs=b.style;
-      bs.height=bs.maxHeight=paddedHeight+'px';
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:1266]: That extracted bit that now appears elsewhere.

File: web/source/osk/emptyView.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+/// <reference path="keyboardView.interface.ts" />
+
+namespace com.keyman.osk {
+  export class EmptyView implements KeyboardView {
+    readonly element: HTMLDivElement;
```
Comment (jahorton) [inline web/source/osk/emptyView.ts:5]: I'm _fairly_ certain KMW currently keeps the OSK hidden whenever this is the active "inner frame" type.  But... it _does_ bother to construct it, so it gets its own proper class.

File: common/core/web/keyboard-processor/src/keyboards/keyboard.ts [hunk 9]
```diff
@@ -98,24 +98,27 @@ namespace com.keyman.keyboards {
...
-     * Reference: https://help.keyman.com/developer/language/reference/kmw_helptext
+     * Reference: https://help.keyman.com/developer/language/reference/kmw_helptext,
+     *            https://help.keyman.com/developer/language/reference/kmw_helpfile
      */
     get helpText(): string {
       return this.scriptObject['KH'];
     }
 
-    get hasHelpHTML(): boolean {
+    get hasScript(): boolean {
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/src/keyboards/keyboard.ts:110]: Please add a comment, similar to the one above for `helpText`, referencing kmw_embedjs

File: common/core/web/keyboard-processor/src/keyboards/keyboard.ts (hunk 1 above)
Comment (mcdurdin) [reply common/core/web/keyboard-processor/src/keyboards/keyboard.ts:117]: How could we chart a way forward to using shadow dom? What sort of changes would we need to make to keyboards that use kmw_embedjs? (sil_euro_latin, Chinese, Japanese, Korean, perhaps others).

File: web/source/osk/visualKeyboard.ts [hunk 10]
```diff
@@ -1527,11 +1523,11 @@ namespace com.keyman.osk {
      *  @param  {Object}            PKbd            the keyboard object to be displayed
      *  @param  {string=}           argFormFactor   layout form factor, defaulting to 'desktop'
      *  @param  {(string|number)=}  argLayerId      name or index of layer to show, defaulting to 'default'
-     *  @param  {Object}            host            KeymanWeb's active OSKManager instance 
+     *  @param  {Object}            height          Target height for the rendered keyboard
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1526]:
```suggestion
     *  @param  {number}            height          Target height for the rendered keyboard 
```

Although we are trying to move types out of jsdoc and into the function signature I suppose...

**Review by mcdurdin — COMMENTED** [review]
There are a fair few architectural changes here so solid testing on iOS, Android and web (desktop + touch modes) will be needed.

Test in particular, as these all make use of the `&kmw_embedjs` property:
* sil_euro_latin
* https://keyman.com/keyboards/tibetan_ewts_to_unicode
* https://keyman.com/keyboards/japanese
* https://keyman.com/keyboards/chinese
* https://keyman.com/keyboards/korean_rr?bcp47=ko 

Apart from sil_euro_latin, these are desktop-web only.

**Comment by jahorton** [comment]
> There are a fair few architectural changes here so solid testing on iOS, Android and web (desktop + touch modes) will be needed.
> 
> Test in particular, as these all make use of the `&kmw_embedjs` property:
> 
>     * sil_euro_latin
> 
>     * https://keyman.com/keyboards/tibetan_ewts_to_unicode
> 
>     * https://keyman.com/keyboards/japanese
> 
>     * https://keyman.com/keyboards/chinese
> 
>     * https://keyman.com/keyboards/korean_rr?bcp47=ko
> 
> 
> Apart from sil_euro_latin, these are desktop-web only.

I'd like to point out that the latter three do have some known pre-existing issues - especially `japanese`, which shows up in keymanweb.com's Sentry logs.

**Comment by mcdurdin** [comment]
> I'd like to point out that the latter three do have some known pre-existing issues - especially `japanese`, which shows up in keymanweb.com's Sentry logs.

FWIW, these _did_ work without errors on KeymanWeb 2.0, so those should be considered regressions.

File: common/core/web/keyboard-processor/src/keyboards/keyboard.ts (hunk 1 above)
Comment (jahorton) [reply common/core/web/keyboard-processor/src/keyboards/keyboard.ts:117]: Chinese and Japanese are deprecated, .js-only keyboards so far as I can tell.  Neither defines a `KHF` function.  `korean_rr` is not deprecated, and while it does use `kmw_embedjs`, it doesn't actually use `KHF` either.  Help text is available, but it doesn't render a "help page" like `sil_eurolatin` does.

`tibetan_ewts_to_unicode` (also in legacy) is the only other keyboard that defines a `KHF`.

My proposition:

We pass `KHF` a `ShadowRoot` (when possible) specific to the keyboard, rather than the OSK's root `_Box`.

- If necessary (for _old_ browsers / pre-Chromium Edge), JS code can detect whether or not it received an actual shadow root:  https://developer.mozilla.org/en-US/docs/Web/API/Node/getRootNode
- From there:
    1. Build `<style>` element
    2. If ShadowDOM available, insert to shadow root
    3. Else, insert to document (also needed for compat with older KMW versions)
    4. Rinse & repeat with other components.

If the keyboard is able to utilize Shadow DOM stuff, great!  If not... I think we kind of have to support the old, non-shadowed functionality for now - both in keyboards and in the engine.  I don't see any reliable way to shoe-horn non-shadow-aware JS into Shadow DOM stuff.

Comment (mcdurdin) [reply common/core/web/keyboard-processor/src/keyboards/keyboard.ts:117]: > Chinese and Japanese are deprecated

Legacy, not deprecated. Big difference.

**Comment by jahorton** [comment]
> > I'd like to point out that the latter three do have some known pre-existing issues - especially `japanese`, which shows up in keymanweb.com's Sentry logs.
> 
> FWIW, these _did_ work without errors on KeymanWeb 2.0, so those should be considered regressions.

Are we sure about that?  We didn't exactly have Sentry reporting then, and the biggest offender - `japanese` - generally functions correctly.  We know that it does produce errors, but not exactly when and why.  The quickest repro I can figure out:

https://sentry.keyman.com/share/issue/e5dc7d5d8a41448d8acc77b68074b594/

I raised this event by attempting to use the `japanese` keyboard while emulating a mobile device in Chrome.  It occurred after typing a single keystroke.  Now, I'd like to point out one extremely important detail here:  The `japanese` keyboard was only ever designed to work on `desktop` form factors.  Heck, it's supposed to be deprecated... so it's a bit interesting that it's even available through keymanweb.com.  (I'm guessing _this_ is because it's the only keyboard for the language?)

**Comment by mcdurdin** [comment]
> Heck, it's supposed to be deprecated...

It's not deprecated. It's legacy.

**Comment by mcdurdin** [comment]
> > Heck, it's supposed to be deprecated...
> 
> It's not deprecated. It's legacy.

What's more, that doesn't even matter. We have a [principle](https://github.com/keymanapp/keyman/wiki/Principles-of-Keyman-Code-Changes#2-engine-backward-compatibility-a-compiled-keyboard-or-model-compiled-in-an-earlier-version-must-work-unmodified-in-a-later-version) of making sure existing keyboards continue to work, even if they are deprecated.

**Comment by jahorton** [comment]
> What's more, that doesn't even matter. We have a [principle](https://github.com/keymanapp/keyman/wiki/Principles-of-Keyman-Code-Changes#2-engine-backward-compatibility-a-compiled-keyboard-or-model-compiled-in-an-earlier-version-must-work-unmodified-in-a-later-version) of making sure existing keyboards continue to work, even if they are deprecated.

Not like I ever said it wasn't.  That said:

> FWIW, these did work without errors on KeymanWeb 2.0, so those should be considered regressions.

>  Now, I'd like to point out one extremely important detail here: The japanese keyboard was only ever designed to work on desktop form factors.

I'm not 100% that there are no lingering regressions for the CJK keyboards, but the one way I currently know to repro the issues we're getting through Sentry relates to an aspect that is _not_ a regression.  Those keyboards were simply never designed for touch, and this PR isn't the place to attempt a remedy.

**Comment by mcdurdin** [comment]
> Those keyboards were simply never designed for touch, and this PR isn't the place to attempt a remedy.

Agreed. Which I said in the original test notes:

> Apart from sil_euro_latin, these are desktop-web only.

**Comment by jahorton** [comment]
Pinging @keymanapp/testers:  this is current top-level `web` PR.

**Comment by MakaraSok** [comment]
### Explicit test list

Run this suite on the following platform / browser combinations:

- [x] Windows / Firefox - **PASSED**
- [x] Windows / Chrome - **PASSED**
- [x] macOS / Safari - **PASSED**
<details>
<summary><strong>Per-platform suite and template</strong> (desktop form-factor)</summary>

```
- [ ] `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- [ ] `LATIN.ROBUST.DESKTOP.1`:  physically type `=` then `e`.
    - Expected result: `ə`  
- [ ] `LATIN.ROBUST.DESKTOP.2`:  physically type `m`.
    - Expected result: `m`
- [ ] `LATIN.SHORTCUT`:  click on the "help-text OSK"'s `ə` button
    - Expected result: `ə`
    - Refer to the following image for the "button": 
        <img width="139" alt="image" src="https://user-images.githubusercontent.com/25213402/126125311-6c95a743-a87f-4bdf-bc24-090a3c449fa1.png">
- [ ] `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- [ ] `KHMER.ROBUST.1`:  physically type a `ស`.
```

- `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- `LATIN.ROBUST.DESKTOP.1`:  physically type `=` then `e`.
    - Expected result: `ə`  
- `LATIN.ROBUST.DESKTOP.2`:  physically type `m`.
    - Expected result: `m`
- `LATIN.SHORTCUT`:  click on the "help-text OSK"'s `ə` button
    - Expected result: `ə`
    - Refer to the following image for the "button": 
        image
- `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- `KHMER.ROBUST.1`:  physically type a `ស`.
</details>

Run _this_ suite on the following platform / browser combinations

- [x] macOS / Chrome emulation / iPhone X - **PASSED**
- [x] Android / Keyman app - **FAILED**: The keyboard is flickering and not usable with a message read "Fatal keyboard error with sil_euro_latin:sil_euro_latin for language. Loading default keyboard." 


<details>
<summary><strong>Per-platform suite and template</strong> (mobile form-factors)</summary>

```
- [ ] `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- [ ] `LATIN.ROBUST.MOBILE.1`:  use the OSK to type `=` (numeric layer) then `e`.
    - Expected result: `=e`
- [ ] `LATIN.ROBUST.MOBILE.2`:  use a longpress on `e` to type `ə`.  (Second-to-last subkey)
    - Expected result: `ə`
- [ ] `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- [ ] `KHMER.ROBUST.1`:  use the OSK to type a `ស`.
```

- `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- `LATIN.ROBUST.MOBILE.1`:  use the OSK to type `=` (numeric layer) then `e`.
    - Expected result: `=e`
- `LATIN.ROBUST.MOBILE.2`:  use a longpress on `e` to type `ə`.  (Second-to-last subkey)
    - Expected result: `ə`
- `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- `KHMER.ROBUST.1`:  use the OSK to type a `ស`.

</details>

**Comment by jahorton** [comment]
So, for Android:

> FAILED: The keyboard is flickering and not usable with a message read "Fatal keyboard error with sil_euro_latin:sil_euro_latin for language. Loading default keyboard."

... how did I not catch that?  Anyway, that's easy enough to repro; fixing it now.

------

Also unchecked the box above; not only did the "test" fail, it was a precondition for the actual tests to be run, none of which did:

**Android / Keyman app**

- [ ] `LATIN.LOAD`:  load the `sil_euro_latin` keyboard; verify everything looks correct
- [ ] `LATIN.ROBUST.MOBILE.1`:  use the OSK to type `=` (numeric layer) then `e`.
    - Expected result: `=e`
- [ ] `LATIN.ROBUST.MOBILE.2`:  use a longpress on `e` to type `ə`.  (Second-to-last subkey)
    - Expected result: `ə`
- [ ] `KHMER.LOAD`: load the `khmer_angkor` keyboard; verify everything looks correct.
- [ ] `KHMER.ROBUST.1`:  use the OSK to type a `ស`.

**Comment by jahorton** [comment]
OK, should be ready for a re-test.  Looping error should be dead now.

**Comment by MakaraSok** [comment]
@jahorton It looks like the build fail for web. Should I wait for the build to get through or retest now?

**Comment by jahorton** [comment]
It's fine now.  Just needed a re-run; one of the test configs failed to launch properly that time.

We should be go for a re-test now.

**Comment by MakaraSok** [comment]
## Android / Keyman app - **ALL PASSED**


- [x] `LATIN.LOAD`: **PASSED** - `sil_euro_latin` keyboard loaded without the eerie error and everything looks correct
- [x] `LATIN.ROBUST.MOBILE.1`: **PASSED** - typing `=` (numeric layer) then `e` outputs the expected result: `=e`.
- [x] `LATIN.ROBUST.MOBILE.2`: **PASSED** - longpressing on `e` to type `ə` outputs the expected result: `ə`.
- [x] `KHMER.LOAD`: **PASSED** - `khmer_angkor` keyboard loaded and everything looks correct.
- [x] `KHMER.ROBUST.1`: **PASSED** - typing a `ស` outputs the expected character.

**Review by mcdurdin — APPROVED** [review]
LGTM. Only minor nits, requests to add/update documentation comments from earlier review to address as well.

File: web/source/kmwembedded.ts [hunk 11]
```diff
@@ -255,7 +255,11 @@ namespace com.keyman.text {
    * correctOSKTextSize handles rotation event -- currently rebuilds keyboard and adjusts font sizes
    */
   keymanweb['correctOSKTextSize']=function() {
-    if(osk && osk.vkbd && osk.vkbd.adjustHeights(osk)) {
+    let osk = keymanweb.osk;
+    if(osk && osk.vkbd && osk.vkbd.adjustHeights(osk.getKeyboardHeight())) {
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwembedded.ts:259]: If our current TS version supports it, absolutely.  I really do like that about Swift... I just didn't think to check if TS actually had it.

[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwembedded.ts:259]: ... and it is!  Creates the following _super-elegant_ result when compiled:

```
if ((_a = osk === null || osk === void 0 ? void 0 : osk.vkbd) === null || _a === void 0 ? void 0 : _a.adjustHeights(osk.getKeyboardHeight())) {
```

And today I learned about `void 0` and that it's _more_ backward-compatible than `undefined` is!

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.88-alpha](https://keyman.com/downloads/releases/alpha/15.0.88)
